### PR TITLE
ESP-IDF 5 MCPWM Implementation Updates for Wall-E

### PR DESCRIPTION
**Pull request by SuperChamp234**
This PR contains the changes for using the new MCPWM implementation as per ESP-IDF 5.1.

With this PR, Wall-E will use the new SRA Board component including the MCPWM changes for ESP-IDF 5.1. There is also an update in the line following code due to PCB changes.

Please review and let me know your suggestions.

### CODE REVIEW

File: 7_self_balancing/main/tuning_websocket_server.c [hunk 1]
```diff
@@ -82,6 +82,20 @@ void websocket_callback(uint8_t num, WEBSOCKET_TYPE_t type, char *msg, uint64_t
     }
 }
 
+void plot_graph_task(void* pvParameters){
+
+    //try to read from the queue, if it is not empty and we successfully receive a value then call plot_graph function
+    //if the queue is empty then wait for 10ms and try again
+    plot_graph_data_t* pg_data;
```
Comment (RISHI27-dot) [inline 7_self_balancing/main/tuning_websocket_server.c:90]: Change to `plot_graph_data_t *pg_data`

File: 7_self_balancing/main/include/tuning_websocket_server.h [hunk 2]
```diff
@@ -39,6 +39,16 @@ typedef struct pid_const
...
 
+typedef struct plot_graph_data {
+    float p_term;
+    float d_term;
+    float i_term;
+    float pitch_corr;
+    float pitch_err;
+} plot_graph_data_t;
+
+void plot_graph_task(void* pvParameters);
```
[the remark this answers is not in the document]
Comment (SuperChamp234) [reply 7_self_balancing/main/include/tuning_websocket_server.h]: Ok

File: 7_self_balancing/main/self_balancing.c [hunk 3]
```diff
@@ -20,6 +20,9 @@
 float forward_offset = 2.51f;
 float forward_buffer = 3.1f;
 */
+plot_graph_data_t pg_data;
+plot_graph_data_t *pg_data_handle;
```
[the remark this answers is not in the document]
Comment (SuperChamp234) [reply 7_self_balancing/main/self_balancing.c]: Done

File: 7_self_balancing/main/tuning_websocket_server.c [hunk 4]
```diff
@@ -82,6 +82,20 @@ void websocket_callback(uint8_t num, WEBSOCKET_TYPE_t type, char *msg, uint64_t
     }
 }
 
+void plot_graph_task(void* pvParameters){
```
[the remark this answers is not in the document]
Comment (SuperChamp234) [reply 7_self_balancing/main/tuning_websocket_server.c]: Okay

File: 7_self_balancing/main/tuning_websocket_server.c (hunk 1 above)
Comment (SuperChamp234) [reply 7_self_balancing/main/tuning_websocket_server.c:90]: Done

File: 7_self_balancing/main/tuning_websocket_server.c [hunk 5]
```diff
@@ -82,6 +82,20 @@ void websocket_callback(uint8_t num, WEBSOCKET_TYPE_t type, char *msg, uint64_t
...
+
+    //try to read from the queue, if it is not empty and we successfully receive a value then call plot_graph function
+    //if the queue is empty then wait for 10ms and try again
```
[the remark this answers is not in the document]
Comment (SuperChamp234) [reply 7_self_balancing/main/tuning_websocket_server.c]: Okay

File: 6_line_following/main/line_following.c
[the remark this answers is not in the document]
Comment (SuperChamp234) [reply 6_line_following/main/line_following.c:1]: I have removed them now.

File: 7_self_balancing/main/self_balancing.c [hunk 6]
```diff
@@ -57,7 +59,16 @@ void calculate_motor_command(const float pitch_error, float *motor_cmd)
...
 
-    plot_graph(P_term, D_term, I_term, pitch_correction, pitch_error);
+	pg_data.p_term = P_term;
+	pg_data.d_term = D_term;
+	pg_data.i_term = I_term;
+	pg_data.pitch_corr = pitch_correction;
+	pg_data.pitch_err = pitch_error;
+	
+	plot_graph_data_t *pg_data_handle = &pg_data;
+	//ESP_LOGI("debug", "P_term: %f ::  D_term: %f  :: I_term: %f :: Pitch_correction: %0.2f :: Pitch_error: %0.2f", P_term, D_term, I_term, pitch_correction, pitch_error);
```
[the remark this answers is not in the document]
Comment (SuperChamp234) [reply 7_self_balancing/main/self_balancing.c]: Done.

File: 7_self_balancing/main/self_balancing.c [hunk 7]
```diff
@@ -170,9 +182,13 @@ void balance_task(void *arg)
 
 void app_main()
 {
-  // Starts tuning server for wireless control
+	//init a queue to send pointers to plot_graph_data_t structs
```
[the remark this answers is not in the document]
Comment (SuperChamp234) [reply 7_self_balancing/main/self_balancing.c:184]: Done

File: 7_self_balancing/main/tuning_websocket_server.c [hunk 8]
```diff
@@ -82,6 +82,21 @@ void websocket_callback(uint8_t num, WEBSOCKET_TYPE_t type, char *msg, uint64_t
     }
 }
 
+void plot_graph_task(void* plot_data_queue){
```
Comment (RISHI27-dot) [inline 7_self_balancing/main/tuning_websocket_server.c:85]: change to `void *plot_data_queue`

Comment (SuperChamp234) [reply 7_self_balancing/main/tuning_websocket_server.c]: Okay

File: 7_self_balancing/main/self_balancing.c [hunk 9]
```diff
@@ -170,9 +181,13 @@ void balance_task(void *arg)
 
 void app_main()
 {
-  // Starts tuning server for wireless control
+	// Create a queue to send PID values to plot graph task
+	plot_graph_queue = xQueueCreate(10, sizeof( &pg_data));
```
Comment (RISHI27-dot) [inline 7_self_balancing/main/self_balancing.c:185]: remove space `plot_graph_queue = xQueueCreate(10, sizeof( &pg_data));`

File: 7_self_balancing/main/self_balancing.c [hunk 10]
```diff
@@ -57,7 +59,15 @@ void calculate_motor_command(const float pitch_error, float *motor_cmd)
...
-    plot_graph(P_term, D_term, I_term, pitch_correction, pitch_error);
+	pg_data.p_term = P_term;
+	pg_data.d_term = D_term;
+	pg_data.i_term = I_term;
+	pg_data.pitch_corr = pitch_correction;
+	pg_data.pitch_err = pitch_error;
+	
+	plot_graph_data_t *pg_data_handle = &pg_data;
+
+	xQueueSend(plot_graph_queue, (void *) &pg_data_handle, (TickType_t) 0);
```
Comment (RISHI27-dot) [inline 7_self_balancing/main/self_balancing.c:70]: Remove extra space and please follow good coding standards 
`xQueueSend(plot_graph_queue, (void *)&pg_data_handle, (TickType_t) 0);`

File: 7_self_balancing/main/tuning_websocket_server.c [hunk 11]
```diff
@@ -82,6 +82,21 @@ void websocket_callback(uint8_t num, WEBSOCKET_TYPE_t type, char *msg, uint64_t
...
 }
 
+void plot_graph_task(void *plot_data_queue){
+    /*
+        Try to read from the queue, if it is not empty and we successfully receive a value then call plot_graph function
+        If the queue is empty then wait for 10ms and try again 
+    */
+    plot_graph_data_t *pg_data;
+    while(1){
+        if(xQueueReceive((QueueHandle_t) plot_data_queue, &pg_data, (TickType_t) 10) == pdPASS){
```
Comment (RISHI27-dot) [inline 7_self_balancing/main/tuning_websocket_server.c:92]: Change to `xQueueReceive((QueueHandle_t)plot_data_queue, &pg_data, (TickType_t)10) == pdPASS`

Comment (SuperChamp234) [reply 7_self_balancing/main/tuning_websocket_server.c]: Yes updated

File: 7_self_balancing/main/self_balancing.c (hunk 9 above)
Comment (SuperChamp234) [reply 7_self_balancing/main/self_balancing.c]: Yes Updated

File: 7_self_balancing/main/self_balancing.c (hunk 10 above)
Comment (SuperChamp234) [reply 7_self_balancing/main/self_balancing.c]: Yes, Updated

File: 4_switch_controlled_motor_normal/main/switch_control_motor_normal.c
Comment (RISHI27-dot) [inline 4_switch_controlled_motor_normal/main/switch_control_motor_normal.c:1]: LGTM!!

File: 5_PWM/main/pwm_test.c
Comment (RISHI27-dot) [inline 5_PWM/main/pwm_test.c:1]: LGTM!!

File: 6_line_following/main/line_following.c
Comment (RISHI27-dot) [inline 6_line_following/main/line_following.c:1]: LGTM!!

File: 7_self_balancing/main/include/tuning_websocket_server.h
Comment (RISHI27-dot) [inline 7_self_balancing/main/include/tuning_websocket_server.h:1]: LGTM!!

File: 4_switch_controlled_motor_normal/main/switch_control_motor_normal.c
Comment (RISHI27-dot) [inline 4_switch_controlled_motor_normal/main/switch_control_motor_normal.c:1]: LGTM!!

File: 5_PWM/main/pwm_test.c
Comment (RISHI27-dot) [inline 5_PWM/main/pwm_test.c:1]: LGTM!!

File: 6_line_following/main/line_following.c
Comment (RISHI27-dot) [inline 6_line_following/main/line_following.c:1]: LGTM!!

File: 7_self_balancing/main/self_balancing.c
Comment (RISHI27-dot) [inline 7_self_balancing/main/self_balancing.c:1]: Looks Good !!

File: 7_self_balancing/main/tuning_websocket_server.c
Comment (RISHI27-dot) [inline 7_self_balancing/main/tuning_websocket_server.c:1]: LGTM!!

**Review by RISHI27-dot — APPROVED** [review]

**Pull request merged**